src/rules/noRestrictedGlobalsRule.ts

```diff
@@ -0,0 +1,97 @@
+    /* tslint:disable:object-literal-sort-keys */
+    public static metadata: Lint.IRuleMetadata = {
+        ruleName: "no-restricted-globals",
+        description: "Disallow specific global variables.",
+        descriptionDetails: Lint.Utils.dedent`
+            Disallowing usage of specific global variables can be useful if you want to allow
+            a set of global variables by enabling an environment, but still want to disallow
+            some of those. For instance, early Internet Explorer versions exposed the current
+            DOM event as a global variable event, but using this variable has been considered
+            as a bad practice for a long time. Restricting this will make sure this variable
+            isn’t used in browser code.
+        `,
+        optionsDescription: "This rule takes a list of strings, where each string is a global to be restricted.",
+        options: {
+            type: "list",
+            items: {type: "string"},
+        },
+        optionExamples: [
+            [
+                true,
+                "name",
+                "length",
+                "event",
+            ],
+        ],
+        type: "functionality",
+        typescriptOnly: false,
+        requiresTypeInfo: true,
+    };
+    /* tslint:enable:object-literal-sort-keys */
+
+    public static FAILURE_STRING(name: string) {
+        return `Unexpected global variable '${name}'. Use local parameter instead.`;
+    }
+
+    public applyWithProgram(sourceFile: ts.SourceFile, program: ts.Program): Lint.RuleFailure[] {
+        const bannedGlobals = new Set(this.ruleArguments as string[]);
+        return this.applyWithFunction(sourceFile, walk, bannedGlobals, program.getTypeChecker());
+    }
+}
+
+function walk(ctx: Lint.WalkContext<Set<string>>, checker: ts.TypeChecker): void {
+    return ts.forEachChild(ctx.sourceFile, function recur(node: ts.Node): void {
+        switch (node.kind) {
+            case ts.SyntaxKind.TypeReference:
+                // Ignore types.
+                return;
+            case ts.SyntaxKind.PropertyAccessExpression:
+                // Ignore `y` in `x.y`, but recurse to `x`.
+                return recur((node as ts.PropertyAccessExpression).expression);
+            case ts.SyntaxKind.Identifier:
+                return checkIdentifier(node as ts.Identifier);
+            default:
+                return ts.forEachChild(node, recur);
+        }
+    });
+
+    function checkIdentifier(node: ts.Identifier): void {
+        if (!ctx.options.has(node.text)) {
+            return;
+        }
+
+        const symbol = checker.getSymbolAtLocation(node);
+        const declarations = symbol === undefined ? undefined : symbol.declarations;
+        if (declarations === undefined || declarations.length === 0) {
+            return;
+        }
+
+        const declaredInLibDom = declarations.some((decl) => decl.getSourceFile().fileName.endsWith(".d.ts"));
```

Comment:
What if you declare something in an `externals.d.ts` file, or define your own `lib.dom.d.ts` with either that name or `my-lib.d.ts`?

Or what about a project that uses `namespace`s and has a `const name = "HA!";` somewhere?

Maybe the rule should get the list of files TS is compiling with _(can we do that?)_ along with its compiler options, and add `lib.d.ts`-like files as per `lib` and `--noLib` compiler settings?
Maybe the rule should keep a default list of regular expressions to match names against, like `lib*.d.ts`?

Reply:
IMO, custom definition files should be considered the same as official `lib.dom.d.ts`. There is no need to get compiler options.